flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractKeyedStateBackend.java

```diff
@@ -201,37 +199,34 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
 	 */
 	@Override
 	public TypeSerializer<K> getKeySerializer() {
-		return keySerializerProvider.currentSchemaSerializer();
+		return keyContext.getCurrentKeySerializer();
 	}
 
 	/**
 	 * @see KeyedStateBackend
 	 */
 	@Override
 	public K getCurrentKey() {
-		return currentKey;
+		return this.keyContext.getCurrentKey();
 	}
 
 	/**
 	 * @see KeyedStateBackend
 	 */
-	@Override
 	public int getCurrentKeyGroupIndex() {
-		return currentKeyGroup;
+		return this.keyContext.getCurrentKeyGroupIndex();
 	}
 
 	/**
 	 * @see KeyedStateBackend
 	 */
-	@Override
 	public int getNumberOfKeyGroups() {
 		return numberOfKeyGroups;
```

Comment:
Why can't we also just depend on the `keyContext` for the number of key groups and key group range also?

Reply:
Semantically calling `keyContext.getNumberOfKeyGroups` is equal to current implementation, but since `numberOfKeyGroups` is final, recording it in a local field could assure no method invocation happens during the `getNumberOfKeyGroups` call. I'm not sure whether the JIT compiler will optimize this, but I guess it's a Hotspot-dependent behavior, so for all final fields in `keyContext` I kept a local field in `AbstractKeyedStateBackend`.